# Phase 1 — AI Agent Conversation Quality Scorer
## Domain Research + Dataset + EDA + Baselines

**Task framing.** An AI-agent *conversation quality scorer* needs, at minimum, to tell when an
agent's answer is **ungrounded / hallucinated** vs. faithful to the source it was given. We treat
this as binary grounding-faithfulness classification and build on the canonical public benchmark.

**Dataset — HaluEval-QA** (Li et al., *HaluEval*, EMNLP 2023, RUCAIBox). 10,000 HotpotQA-derived
items; each item carries a `knowledge` passage, a `question`, a grounded `right_answer`, and a
ChatGPT-generated `hallucinated_answer`. Expanding each item into its two answers yields a
**perfectly balanced 20,000-sample** binary set (0 = grounded, 1 = hallucinated). It is the right
Phase-1 choice because it has *named* fields (not anonymized PCA), an established leaderboard, and a
clean protocol.

**Published benchmark to beat / contextualize.** The HaluEval paper reports **ChatGPT zero-shot at
62.59% accuracy** on the QA hallucination-classification task — i.e. frontier-2023 instruction
models were barely above chance at *spotting* hallucinations they themselves help generate.

**Primary metric — macro-F1.** The task is balanced (50/50) but we care equally about both the
*grounded* and *hallucinated* classes (a quality scorer that only catches one is useless), and the
HaluEval protocol scores accuracy on a balanced set, so macro-F1 collapses to ≈accuracy here while
staying robust if later splits drift from balance. We track accuracy, balanced-acc, per-class
precision/recall, and ROC-AUC as secondaries. **All leaderboards in this project rank by macro-F1.**

**The Phase-1 question.** Before building anything sophisticated, what is the *honest* floor? In
particular — is HaluEval-QA actually testing grounding, or is it leaking a cheap surface shortcut?
Recent work (*"The Illusion of Progress: Re-evaluating Hallucination Detection in LLMs"*, arXiv
2508.08285, 2025) warns that hallucinated text is systematically **longer**, and that detectors
silently exploit length rather than meaning. We test that here.

In [1]:
import json, os, re, time
from pathlib import Path
import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt

from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GroupShuffleSplit
from sklearn.metrics import (f1_score, accuracy_score, balanced_accuracy_score,
                             precision_score, recall_score, roc_auc_score,
                             confusion_matrix)
import sklearn
print("sklearn", sklearn.__version__, "| numpy", np.__version__, "| pandas", pd.__version__)

RANDOM_STATE = 42

def find_root():
    p = Path.cwd()
    for cand in [p, *p.parents]:
        if (cand / "data" / "raw" / "qa_data.json").exists():
            return cand
    return p

ROOT = find_root()
RESULTS = ROOT / "results"; RESULTS.mkdir(exist_ok=True)
plt.rcParams.update({"figure.dpi": 110, "axes.grid": True, "grid.alpha": 0.3})
print("project root:", ROOT)

sklearn 1.8.0 | numpy 2.4.2 | pandas 3.0.1
project root: /Users/anthonyrodrigues/Desktop/YC-Portfolio-Projects/AI-Agent-Conversation-Quality-Scorer


## 1. Load & reshape

The raw file is JSONL: one HotpotQA item per line. We **expand** each item into two labeled rows
(grounded → 0, hallucinated → 1) that share the same `qid`, `question`, and `knowledge`. The shared
`qid` is what lets us split *by item* later so the two answers of one item can never straddle the
train/test boundary (no question leakage).

In [2]:
raw_path = ROOT / "data" / "raw" / "qa_data.json"
records = [json.loads(l) for l in raw_path.read_text().splitlines() if l.strip()]
print("raw HaluEval-QA items:", len(records))

long = []
for qid, r in enumerate(records):
    long.append({"qid": qid, "knowledge": r["knowledge"], "question": r["question"],
                 "answer": r["right_answer"], "label": 0})        # grounded
    long.append({"qid": qid, "knowledge": r["knowledge"], "question": r["question"],
                 "answer": r["hallucinated_answer"], "label": 1}) # hallucinated
df = pd.DataFrame(long)
print("expanded samples:", df.shape)
print("class balance:", df.label.value_counts(normalize=True).round(4).to_dict())
display(df.head(4))

raw HaluEval-QA items: 10000


expanded samples: (20000, 5)
class balance: {0: 0.5, 1: 0.5}


,qid,knowledge,question,answer,label
0,0,Arthur's Magazine (1844–1846) was an American ...,Which magazine was started first Arthur's Maga...,Arthur's Magazine,0
1,0,Arthur's Magazine (1844–1846) was an American ...,Which magazine was started first Arthur's Maga...,First for Women was started first.,1
2,1,The Oberoi family is an Indian family that is ...,The Oberoi family is part of a hotel company t...,Delhi,0
3,1,The Oberoi family is an Indian family that is ...,The Oberoi family is part of a hotel company t...,The Oberoi family's hotel company is based in ...,1


## 2. Data-quality investigation

Before modeling: missingness, exact-duplicate answers, and degenerate (empty) text. Systematic
problems here would invalidate any downstream metric.

In [3]:
qc = {
    "n_samples": int(len(df)),
    "n_items": int(df.qid.nunique()),
    "missing_answer": int(df.answer.isna().sum()),
    "empty_answer": int((df.answer.str.strip().str.len() == 0).sum()),
    "missing_knowledge": int(df.knowledge.isna().sum()),
    "dup_full_rows": int(df.duplicated(subset=["qid","answer","label"]).sum()),
    "grounded==hallucinated (collisions)": int(
        sum(r["right_answer"].strip().lower() == r["hallucinated_answer"].strip().lower()
            for r in records)),
}
for k, v in qc.items():
    print(f"{k:42s}: {v}")

n_samples                                 : 20000
n_items                                   : 10000
missing_answer                            : 0
empty_answer                              : 0
missing_knowledge                         : 0
dup_full_rows                             : 0
grounded==hallucinated (collisions)       : 0


## 3. The length-asymmetry probe

Grounded answers in HaluEval-QA are HotpotQA gold strings — usually a single **entity** ("Arthur's
Magazine"). Hallucinated answers are ChatGPT-written **full sentences** ("First for Women was started
first."). If that style gap is large, a model can "detect hallucination" by *counting characters* and
never look at the knowledge at all. We quantify the gap, then weaponize it as a baseline.

In [4]:
df["ans_chars"]  = df.answer.str.len()
df["ans_tokens"] = df.answer.str.split().map(len)
df["q_chars"]    = df.question.str.len()
df["k_chars"]    = df.knowledge.str.len()
df["ans_has_period"] = df.answer.str.rstrip().str.endswith((".", "!", "?")).astype(int)

g = df.groupby("label")[["ans_chars","ans_tokens","ans_has_period"]].mean().round(2)
g.index = ["grounded (0)", "hallucinated (1)"]
display(g)
ratio = g.loc["hallucinated (1)","ans_chars"] / g.loc["grounded (0)","ans_chars"]
print(f"hallucinated/grounded mean answer-char ratio: {ratio:.2f}x")
print(f"hallucinated answers ending in punctuation: "
      f"{df[df.label==1].ans_has_period.mean():.1%}  vs grounded: {df[df.label==0].ans_has_period.mean():.1%}")

,ans_chars,ans_tokens,ans_has_period
grounded (0),13.65,2.22,0.01
hallucinated (1),66.22,10.99,0.95


hallucinated/grounded mean answer-char ratio: 4.85x
hallucinated answers ending in punctuation: 95.2%  vs grounded: 1.4%


**Note on which length features can possibly help.** `q_chars` and `k_chars` are *identical* for
the two answers of any given item, so within a matched pair they carry **zero** discriminative signal —
only the *answer* length differs. We therefore expect an answer-length-only model to match the full
4-feature length model. We test exactly that below; if true, it is direct evidence the benchmark's
"signal" is answer surface form, not grounding.

## 4. Grounding-overlap feature (cheap NLI proxy)

The *honest* signal for grounding is: how much of the answer is actually supported by the knowledge?
We use lexical Jaccard — fraction of answer tokens that appear in the knowledge passage. Grounded
entity answers are usually copied from the passage (overlap ≈ 1); a genuine hallucination asserts
tokens not in the source. This is a poor-man's NLI and will be the *most honest* baseline.

In [5]:
STOP = set("a an the of to in on at for and or is was were are be been by with as that this it from".split())
def toks(s):
    return [t for t in re.findall(r"[a-z0-9]+", str(s).lower()) if t not in STOP]
def grounding_overlap(ans, know):
    a = set(toks(ans)); k = set(toks(know))
    return (len(a & k) / len(a)) if a else 0.0

df["ground_overlap"] = [grounding_overlap(a, k) for a, k in zip(df.answer, df.knowledge)]
ov = df.groupby("label")["ground_overlap"].mean().round(3)
print("mean answer→knowledge token overlap   grounded:", ov[0], " hallucinated:", ov[1])
print("(higher overlap = better grounded; gap is the only 'meaning' signal we have pre-modeling)")

mean answer→knowledge token overlap   grounded: 0.936  hallucinated: 0.612
(higher overlap = better grounded; gap is the only 'meaning' signal we have pre-modeling)


## 5. Leak-free split (by item)

`GroupShuffleSplit` on `qid` → 80/20. Both answers of an item land on the same side, so the model
can never memorize an item from its grounded twin.

In [6]:
gss = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=RANDOM_STATE)
tr_idx, te_idx = next(gss.split(df, df.label, groups=df.qid))
train, test = df.iloc[tr_idx].copy(), df.iloc[te_idx].copy()
assert not (set(train.qid) & set(test.qid)), "qid leakage!"
print(f"train: {len(train)} samples / {train.qid.nunique()} items   "
      f"test: {len(test)} samples / {test.qid.nunique()} items")
print("train balance:", train.label.mean().round(3), " test balance:", test.label.mean().round(3))

# persist qid lists so every later phase reuses this EXACT split
json.dump({"train_qids": sorted(map(int, train.qid.unique())),
           "test_qids":  sorted(map(int, test.qid.unique())),
           "random_state": RANDOM_STATE, "test_size": 0.2},
          open(RESULTS / "phase1_split_qids.json", "w"))
print("saved -> results/phase1_split_qids.json")

train: 16000 samples / 8000 items   test: 4000 samples / 2000 items
train balance: 0.5  test balance: 0.5
saved -> results/phase1_split_qids.json


## 6. Baselines

A shared scorer so every row of the leaderboard is computed identically.

In [7]:
def evaluate(name, y_true, y_pred, y_score=None):
    return {
        "model": name,
        "accuracy":        round(accuracy_score(y_true, y_pred), 4),
        "macro_f1":        round(f1_score(y_true, y_pred, average="macro"), 4),
        "balanced_acc":    round(balanced_accuracy_score(y_true, y_pred), 4),
        "precision_hallu": round(precision_score(y_true, y_pred, pos_label=1, zero_division=0), 4),
        "recall_hallu":    round(recall_score(y_true, y_pred, pos_label=1, zero_division=0), 4),
        "roc_auc": round(roc_auc_score(y_true, y_score), 4) if y_score is not None else None,
    }

rows = []

**Baseline 1 — majority class.** The floor. macro-F1 collapses below accuracy because one class is never predicted.

In [8]:
dummy = DummyClassifier(strategy="most_frequent").fit(train[["ans_chars"]], train.label)
pred = dummy.predict(test[["ans_chars"]])
rows.append(evaluate("majority_class", test.label, pred, None))
print(rows[-1])

{'model': 'majority_class', 'accuracy': 0.5, 'macro_f1': 0.3333, 'balanced_acc': 0.5, 'precision_hallu': 0.0, 'recall_hallu': 0.0, 'roc_auc': None}


**Baseline 2 — length-only LogReg (the shortcut probe).** 4 numeric features: answer chars, answer tokens, question chars, knowledge chars. If this is near-perfect, the benchmark is leaking length.

In [9]:
LEN_FEATS = ["ans_chars","ans_tokens","q_chars","k_chars"]
len_pipe = Pipeline([("sc", StandardScaler()),
                     ("lr", LogisticRegression(max_iter=2000, random_state=RANDOM_STATE))])
len_pipe.fit(train[LEN_FEATS], train.label)
proba = len_pipe.predict_proba(test[LEN_FEATS])[:,1]
rows.append(evaluate("length_only_logreg", test.label, (proba>=.5).astype(int), proba))
print(rows[-1])
print("coefs:", dict(zip(LEN_FEATS, len_pipe.named_steps['lr'].coef_[0].round(3))))

{'model': 'length_only_logreg', 'accuracy': 0.9437, 'macro_f1': 0.9437, 'balanced_acc': 0.9437, 'precision_hallu': 0.9521, 'recall_hallu': 0.9345, 'roc_auc': 0.9713}
coefs: {'ans_chars': np.float64(2.222), 'ans_tokens': np.float64(4.565), 'q_chars': np.float64(-0.039), 'k_chars': np.float64(-0.132)}


**Baseline 2b — answer-length ONLY (2 features).** Drops the within-pair-constant question/knowledge lengths. If it matches Baseline 2, the discriminative signal is purely *answer* length.

In [10]:
ALEN = ["ans_chars","ans_tokens"]
alen_pipe = Pipeline([("sc", StandardScaler()),
                      ("lr", LogisticRegression(max_iter=2000, random_state=RANDOM_STATE))])
alen_pipe.fit(train[ALEN], train.label)
proba = alen_pipe.predict_proba(test[ALEN])[:,1]
rows.append(evaluate("answer_length_only_logreg", test.label, (proba>=.5).astype(int), proba))
print(rows[-1])

{'model': 'answer_length_only_logreg', 'accuracy': 0.9435, 'macro_f1': 0.9435, 'balanced_acc': 0.9435, 'precision_hallu': 0.9516, 'recall_hallu': 0.9345, 'roc_auc': 0.9717}


**Baseline 3 — TF-IDF on the answer.** Word 1–2 grams + LogReg. Captures lexical style; expected to be strong but largely redundant with length.

In [11]:
tf_a = Pipeline([("tf", TfidfVectorizer(ngram_range=(1,2), min_df=3, max_features=50000)),
                 ("lr", LogisticRegression(max_iter=2000, C=1.0, random_state=RANDOM_STATE))])
tf_a.fit(train.answer, train.label)
proba = tf_a.predict_proba(test.answer)[:,1]
rows.append(evaluate("tfidf_answer_logreg", test.label, (proba>=.5).astype(int), proba))
print(rows[-1])

{'model': 'tfidf_answer_logreg', 'accuracy': 0.9195, 'macro_f1': 0.9194, 'balanced_acc': 0.9195, 'precision_hallu': 0.952, 'recall_hallu': 0.8835, 'roc_auc': 0.9674}


**Baseline 4 — TF-IDF on question ⊕ answer.** Does adding the question help? The question is *shared* across an item's two answers, so its tokens are non-discriminative noise. Hypothesis: it HURTS.

In [12]:
tr_qa = train.question + " [SEP] " + train.answer
te_qa = test.question  + " [SEP] " + test.answer
tf_qa = Pipeline([("tf", TfidfVectorizer(ngram_range=(1,2), min_df=3, max_features=50000)),
                  ("lr", LogisticRegression(max_iter=2000, C=1.0, random_state=RANDOM_STATE))])
tf_qa.fit(tr_qa, train.label)
proba = tf_qa.predict_proba(te_qa)[:,1]
rows.append(evaluate("tfidf_q_plus_a_logreg", test.label, (proba>=.5).astype(int), proba))
print(rows[-1])

{'model': 'tfidf_q_plus_a_logreg', 'accuracy': 0.8037, 'macro_f1': 0.8035, 'balanced_acc': 0.8037, 'precision_hallu': 0.8275, 'recall_hallu': 0.7675, 'roc_auc': 0.8809}


**Baseline 5 — grounding-overlap threshold.** Predict *hallucination* when the answer's token
overlap with the knowledge is low. Threshold tuned on TRAIN to max macro-F1, then frozen for TEST.
ROC-AUC uses the continuous (1−overlap) score. This is the only baseline that looks at the *source* —
but note its overlap is a *fraction of answer tokens*, so it may be partly length-confounded (short
entity answers trivially have most of their few tokens in the passage). We stress-test that in §7.

In [13]:
ov_tr, ov_te = train.ground_overlap.values, test.ground_overlap.values
best_t, best_f1 = 0.5, -1.0
for t in np.linspace(0, 1, 201):
    f = f1_score(train.label, (ov_tr < t).astype(int), average="macro")
    if f > best_f1: best_f1, best_t = f, t
pred = (ov_te < best_t).astype(int)
rows.append(evaluate("grounding_overlap_threshold", test.label, pred, 1 - ov_te))
print(f"tuned threshold (overlap <): {best_t:.3f}   train macroF1={best_f1:.4f}")
print(rows[-1])

tuned threshold (overlap <): 0.960   train macroF1=0.9256
{'model': 'grounding_overlap_threshold', 'accuracy': 0.9253, 'macro_f1': 0.9252, 'balanced_acc': 0.9252, 'precision_hallu': 0.9364, 'recall_hallu': 0.9125, 'roc_auc': 0.8989}


## 7. The honest floor — a length-matched control

The raw leaderboard rewards any model that exploits the length gap. To see what survives once
length is removed, we build a quick **per-length-bin matched control**: bin TEST answers into 10-char
bins and, within each bin, downsample to equal grounded/hallucinated counts. The pooled set has the
two classes overlapping in length (low KS), so a model can no longer win by reading length alone.
We then re-score *every* baseline on it. This is a lightweight preview of the rigorous matched split
Phase 2 will build.

In [14]:
from scipy.stats import spearmanr, ks_2samp
rho = spearmanr(df.ans_chars, df.ground_overlap).statistic
print(f"Spearman(answer length, grounding overlap) = {rho:.3f}  "
      f"(strongly negative => overlap is largely length in disguise)")

rng = np.random.default_rng(RANDOM_STATE)
test = test.copy()
edges = np.arange(0, int(test.ans_chars.max()) + 11, 10)
test["lbin"] = np.digitize(test.ans_chars, edges)
keep = []
for _, sub in test.groupby("lbin"):
    g0, g1 = sub.index[sub.label == 0], sub.index[sub.label == 1]
    k = min(len(g0), len(g1))
    if k == 0: continue
    keep += list(rng.choice(g0, k, replace=False)) + list(rng.choice(g1, k, replace=False))
matched = test.loc[keep]
ks_raw = ks_2samp(test[test.label==0].ans_chars,    test[test.label==1].ans_chars).statistic
ks_mat = ks_2samp(matched[matched.label==0].ans_chars, matched[matched.label==1].ans_chars).statistic
print(f"length-matched control: n={len(matched)} "
      f"({int((matched.label==0).sum())} grounded / {int((matched.label==1).sum())} hallucinated)")
print(f"answer-length KS between classes:  raw test={ks_raw:.3f}  ->  matched={ks_mat:.3f}")

Spearman(answer length, grounding overlap) = -0.517  (strongly negative => overlap is largely length in disguise)
length-matched control: n=522 (261 grounded / 261 hallucinated)
answer-length KS between classes:  raw test=0.874  ->  matched=0.123


In [15]:
def score_on(name, fr):
    if name == "majority_class":
        p, s = dummy.predict(fr[["ans_chars"]]), None
    elif name == "length_only_logreg":
        s = len_pipe.predict_proba(fr[LEN_FEATS])[:,1];  p = (s>=.5).astype(int)
    elif name == "answer_length_only_logreg":
        s = alen_pipe.predict_proba(fr[ALEN])[:,1];       p = (s>=.5).astype(int)
    elif name == "tfidf_answer_logreg":
        s = tf_a.predict_proba(fr.answer)[:,1];           p = (s>=.5).astype(int)
    elif name == "tfidf_q_plus_a_logreg":
        s = tf_qa.predict_proba(fr.question+" [SEP] "+fr.answer)[:,1]; p = (s>=.5).astype(int)
    elif name == "grounding_overlap_threshold":
        s = 1 - fr.ground_overlap.values; p = (fr.ground_overlap.values < best_t).astype(int)
    return evaluate(name, fr.label, p, s)

rows_matched = [score_on(r["model"], matched) for r in rows]
for r in rows:        r["split"] = "raw"
for r in rows_matched: r["split"] = "length_matched"
print("re-scored", len(rows_matched), "baselines on the matched control")

re-scored 6 baselines on the matched control


## 8. Dual leaderboard — raw vs length-matched

In [16]:
raw_b = pd.DataFrame(rows).sort_values("macro_f1", ascending=False).reset_index(drop=True)
mat_b = pd.DataFrame(rows_matched).sort_values("macro_f1", ascending=False).reset_index(drop=True)
comp = (raw_b[["model","macro_f1"]].rename(columns={"macro_f1":"raw_F1"})
        .merge(mat_b[["model","macro_f1"]].rename(columns={"macro_f1":"matched_F1"}), on="model"))
comp["drop"] = (comp.raw_F1 - comp.matched_F1).round(4)
comp = comp.sort_values("matched_F1", ascending=False).reset_index(drop=True)
comp.index = comp.index + 1
print("RAW leaderboard:"); display(raw_b)
print("\nLENGTH-MATCHED honest leaderboard (ranked by matched macro-F1):"); display(comp)
raw_b.to_csv(RESULTS / "phase1_baselines.csv")
mat_b.to_csv(RESULTS / "phase1_baselines_matched.csv")
comp.to_csv(RESULTS / "phase1_raw_vs_matched.csv")
json.dump(rows + rows_matched, open(RESULTS / "phase1_baselines.json", "w"), indent=2)
print("saved -> phase1_baselines.csv / _matched.csv / phase1_raw_vs_matched.csv / .json")

RAW leaderboard:


,model,accuracy,macro_f1,balanced_acc,precision_hallu,recall_hallu,roc_auc,split
0,length_only_logreg,0.9437,0.9437,0.9437,0.9521,0.9345,0.9713,raw
1,answer_length_only_logreg,0.9435,0.9435,0.9435,0.9516,0.9345,0.9717,raw
2,grounding_overlap_threshold,0.9253,0.9252,0.9252,0.9364,0.9125,0.8989,raw
3,tfidf_answer_logreg,0.9195,0.9194,0.9195,0.9520,0.8835,0.9674,raw
4,tfidf_q_plus_a_logreg,0.8037,0.8035,0.8037,0.8275,0.7675,0.8809,raw
5,majority_class,0.5000,0.3333,0.5000,0.0000,0.0000,NaN,raw



LENGTH-MATCHED honest leaderboard (ranked by matched macro-F1):


,model,raw_F1,matched_F1,drop
1,grounding_overlap_threshold,0.9252,0.9192,0.0060
2,tfidf_answer_logreg,0.9194,0.7131,0.2063
3,tfidf_q_plus_a_logreg,0.8035,0.6915,0.1120
4,length_only_logreg,0.9437,0.6150,0.3287
5,answer_length_only_logreg,0.9435,0.6150,0.3285
6,majority_class,0.3333,0.3333,0.0000


saved -> phase1_baselines.csv / _matched.csv / phase1_raw_vs_matched.csv / .json


## 9. Figures

In [17]:
# Fig 1 — length asymmetry + overlap-vs-length confound
fig, ax = plt.subplots(1, 2, figsize=(11, 4))
bins = np.linspace(0, 160, 60)
ax[0].hist(df[df.label==0].ans_chars, bins=bins, alpha=.7, label="grounded", color="#2a9d8f")
ax[0].hist(df[df.label==1].ans_chars, bins=bins, alpha=.7, label="hallucinated", color="#e76f51")
ax[0].set(title=f"Answer length by class ({ratio:.1f}x asymmetry)", xlabel="answer characters", ylabel="count")
ax[0].legend()
ax[1].scatter(df.ans_chars, df.ground_overlap, s=3, alpha=.15,
              c=df.label.map({0:"#2a9d8f",1:"#e76f51"}))
ax[1].set(title=f"Grounding overlap vs length (Spearman={rho:.2f})",
          xlabel="answer characters", ylabel="answer→knowledge overlap", xlim=(0,160))
plt.tight_layout(); plt.savefig(RESULTS / "phase1_length_and_overlap.png", bbox_inches="tight"); plt.show()

/var/folders/51/mgdjphq11d74ym1_r6mvyx9m0000gn/T/ipykernel_54804/584906777.py:12: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.tight_layout(); plt.savefig(RESULTS / "phase1_length_and_overlap.png", bbox_inches="tight"); plt.show()


In [18]:
# Fig 2 — dual leaderboard: raw vs length-matched
fig, ax = plt.subplots(figsize=(9.5, 5))
order = comp.sort_values("matched_F1")
y = np.arange(len(order)); h = 0.4
ax.barh(y+h/2, order.raw_F1,     h, color="#e9c46a", label="raw split")
ax.barh(y-h/2, order.matched_F1, h, color="#264653", label="length-matched")
ax.set_yticks(y); ax.set_yticklabels(order.model)
for yi, (r, m) in enumerate(zip(order.raw_F1, order.matched_F1)):
    ax.text(r+.005, yi+h/2, f"{r:.3f}", va="center", fontsize=8)
    ax.text(m+.005, yi-h/2, f"{m:.3f}", va="center", fontsize=8)
ax.axvline(0.5, ls=":", color="gray"); ax.text(0.5, len(order)-0.3, "chance", fontsize=8, color="gray")
ax.set(title="Baselines: raw vs length-matched macro-F1\n(collapse toward chance = score was length)",
       xlabel="macro-F1", xlim=(0, 1.02)); ax.legend(loc="lower right")
plt.tight_layout(); plt.savefig(RESULTS / "phase1_baseline_comparison.png", bbox_inches="tight"); plt.show()

/var/folders/51/mgdjphq11d74ym1_r6mvyx9m0000gn/T/ipykernel_54804/2413534870.py:14: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.tight_layout(); plt.savefig(RESULTS / "phase1_baseline_comparison.png", bbox_inches="tight"); plt.show()


In [19]:
# Fig 3 — confusion of the length shortcut: raw vs matched
from itertools import product
def plot_cm(ax, y, p, title):
    cm = confusion_matrix(y, p)
    ax.imshow(cm, cmap="Blues")
    for i, j in product(range(2), range(2)):
        ax.text(j, i, cm[i, j], ha="center", va="center",
                color="white" if cm[i, j] > cm.max()/2 else "black", fontsize=12)
    ax.set(xticks=[0,1], yticks=[0,1], xticklabels=["grnd","hallu"], yticklabels=["grnd","hallu"],
           xlabel="predicted", ylabel="true", title=title)
fig, ax = plt.subplots(1, 2, figsize=(9, 4))
plot_cm(ax[0], test.label, (len_pipe.predict_proba(test[LEN_FEATS])[:,1]>=.5).astype(int),
        "length_only on RAW test")
plot_cm(ax[1], matched.label, (len_pipe.predict_proba(matched[LEN_FEATS])[:,1]>=.5).astype(int),
        "length_only on length-MATCHED")
plt.tight_layout(); plt.savefig(RESULTS / "phase1_confusion_raw_vs_matched.png", bbox_inches="tight"); plt.show()

/var/folders/51/mgdjphq11d74ym1_r6mvyx9m0000gn/T/ipykernel_54804/2454460939.py:16: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.tight_layout(); plt.savefig(RESULTS / "phase1_confusion_raw_vs_matched.png", bbox_inches="tight"); plt.show()


## 10. Persist Phase-1 metrics

In [20]:
matched_floor = mat_b.iloc[0].to_dict()
metrics_path = RESULTS / "metrics.json"
existing = json.load(open(metrics_path)) if metrics_path.exists() else {}
existing["phase1"] = {
    "dataset": "HaluEval-QA (Li et al., EMNLP 2023)",
    "primary_metric": "macro_f1",
    "n_items": int(df.qid.nunique()),
    "n_samples": int(len(df)),
    "class_balance": {str(k): float(v) for k, v in df.label.value_counts(normalize=True).round(4).items()},
    "n_train": int(len(train)), "n_test": int(len(test)),
    "length_asymmetry_ratio": round(float(ratio), 3),
    "overlap_length_spearman": round(float(rho), 3),
    "published_chatgpt_acc": 0.6259,
    "matched_control_n": int(len(matched)),
    "matched_control_ks": round(float(ks_mat), 3),
    "best_raw_macro_f1": float(raw_b.iloc[0].macro_f1),
    "best_matched_macro_f1": float(mat_b.iloc[0].macro_f1),
    "best_matched_model": matched_floor["model"],
    "baselines_raw": rows,
    "baselines_matched": rows_matched,
}
json.dump(existing, open(metrics_path, "w"), indent=2)
print("saved -> results/metrics.json  (keys:", list(existing.keys()), ")")
print(f"best RAW macro-F1     = {raw_b.iloc[0].macro_f1:.4f} ({raw_b.iloc[0].model})")
print(f"best MATCHED macro-F1 = {mat_b.iloc[0].macro_f1:.4f} ({mat_b.iloc[0].model})  <- the honest floor")

saved -> results/metrics.json  (keys: ['phase1'] )
best RAW macro-F1     = 0.9437 (length_only_logreg)
best MATCHED macro-F1 = 0.9192 (grounding_overlap_threshold)  <- the honest floor


## 11. Key findings (Phase 1)

The length-matched control (answer-length KS 0.87 → 0.12) acts as a **truth serum**: it sorts every
cheap feature into "shortcut" vs "real signal," and it overturned one of my own going-in assumptions.

1. **HaluEval-QA is largely solvable by counting characters — and a matched control proves it.**
   Hallucinated answers are ~4.8× longer than grounded entity strings (95% end in punctuation vs 1%
   of grounded), so a length-only LogReg hits **0.944 macro-F1** on the raw split — far above the
   HaluEval paper's ChatGPT zero-shot 0.626 accuracy. But on the length-matched control it **collapses
   to 0.615**, barely above chance. That −0.33 F1 drop *is* the shortcut, made visible. This reproduces
   the *"Illusion of Progress"* (2025) warning: hallucination detectors silently exploit length.
   **Any Phase-2+ "win" on the raw split is meaningless without this control.**
2. **Only the *answer* length matters.** Answer-length-only (2 feat) ≈ the full 4-feature length model
   (both 0.944 raw, both 0.615 matched), because question/knowledge lengths are constant within a
   matched pair. The benchmark's surface "signal" is answer length, nothing else.
3. **I expected grounding-overlap to be length-in-disguise. The control proved me wrong — and that's
   the real finding.** Overlap *is* correlated with length (Spearman = −0.52), so I predicted it would
   collapse with the length models. It did **not**: 0.925 raw → **0.919 matched** (−0.006, essentially
   flat). At *matched* lengths, grounded answers still overlap the source far more than hallucinations
   do — so overlap reads genuine **content grounding**, not length. It is the **honest floor at 0.919**
   and the single cheap baseline worth beating. (Caveat: it leans on HaluEval-QA's construction —
   grounded answers are near-verbatim knowledge spans — so whether it survives on free-form agent
   responses is an open Phase-2 / OOD question.) TF-IDF sits in between (0.919 → 0.713): part style,
   part shortcut.
4. **Adding the question HURTS** (TF-IDF answer 0.919 vs question⊕answer 0.804, raw split). The question
   is shared across an item's two answers, so its tokens are non-discriminative noise — a clean "more
   features ≠ better" result that recurs later in the project.

### Next (Phase 2)
Promote this quick per-bin control to a **rigorous length-matched split** (nearest-length pairing,
report KS), then run 4–6 paradigms — word/char n-grams, sentence-transformer embeddings ± a tree head,
and a zero-shot cross-encoder NLI that has *never* seen HaluEval — on both splits. The bar is now
honest and concrete: **beat grounding-overlap's 0.919 matched macro-F1**, and find out whether a model
that reads *meaning* (NLI / embeddings) can separate hallucinations that even token-overlap misses.